File: services/TodoCsService/Features/TodoList/Infrastructure/TodoModel.cs
using MongoDB.Bson.Serialization;
using TodoCsService.Features.Base;
using TodoCsService.Features.TodoList.Domain.Entity;
using TodoCsService.Features.TodoList.Domain.Value;

namespace TodoCsService.Features.TodoList.Infrastructure;

public record TodoModel(
    string Id,
    int Version,
    string Note,
    DateTime? Due,
    string Status
)
{
    public Todo ToEntity()
    {
        return new Todo(
            TodoId.Parse(Id),
            new AggregateVersion(Version),
            new TodoNote(Note),
            new TodoDue(Due),
            ParseStatus
        );
    }

    public TodoModel Increment() => this with { Version = Version + 1 };

    private TodoStatus ParseStatus => BaseEnum.Parse(Status, TodoStatus.incomplete);

    public static void Register()
    {
        BsonClassMap.RegisterClassMap<TodoModel>(cm =>
        {
            cm.AutoMap();
            cm.MapIdMember(c => c.Id);
        });
    }
}


public static class TodoModelExtensions
{
    public static TodoModel ToModel(this Todo todo)
    {
        return new TodoModel(
            todo.Id.ToString(),
            todo.Version.Value,
            todo.Note.Value,
            todo.Due.Value,
            todo.Status.ToString()
        );
    }
}

File: services/todo-go-service/feature/todo/infrastructure/todo_model.go
package infrastructure

import (
	"time"

	"github.com/sjohnson-pplsi/todo-service/services/todo-go-service/feature/todo/domain/entity"
	"github.com/sjohnson-pplsi/todo-service/services/todo-go-service/feature/todo/domain/value"
)

type TodoModel struct {
	ID      string    `bson:"_id"`
	Version int       `bson:"version"`
	Status  bool      `bson:"status"`
	Note    string    `bson:"note"`
	Due     time.Time `bson:"due"`
}

func (tm *TodoModel) Increment() {
	tm.Version++
}

type todoMapper struct {
}

func (todoMapper) FromModel(m *TodoModel) *entity.Todo {
	return entity.RestoreTodo(
		value.TodoID(m.ID),
		value.Version(m.Version),
		value.TodoStatus(m.Status),
		value.TodoNote(m.Note),
		m.Due,
	)
}

func (todoMapper) ToModel(t *entity.Todo) *TodoModel {
	return &TodoModel{
		ID:      t.ID().String(),
		Version: t.Version().Value(),
		Status:  t.Status().Value(),
		Note:    t.Note().String(),
		Due:     t.Due(),
	}
}
